inst/include/cpp11/r_vector.hpp

```diff
@@ -759,8 +767,25 @@ inline r_vector<T>::r_vector(SEXP&& data, bool is_altrep)
     : cpp11::r_vector<T>(data, is_altrep), capacity_(length_) {}
 
 template <typename T>
-inline r_vector<T>::r_vector(const r_vector& rhs)
-    : cpp11::r_vector<T>(safe[Rf_shallow_duplicate](rhs)), capacity_(rhs.capacity_) {}
+inline r_vector<T>::r_vector(const r_vector& rhs) {
+  // We don't want to just pass through to the read-only constructor because we'd
+  // have to convert to `SEXP` first, which could truncate, and then we'd still have
+  // to shallow duplicate after that to ensure we have a duplicate, which can result in
+  // too many copies (#369).
+  //
+  // Instead we take control of setting all fields to try and only duplicate 1 time.
+  // We try and reclaim unused capacity during the duplication by only reserving up to
+  // the `rhs.length_`. This is nice because if the user returns this object, the
+  // truncation has already been done and they don't have to pay for another allocation.
+  // Importantly, `reserve_data()` always duplicates even if there wasn't extra capacity,
+  // which ensures we have our own copy.
+  data_ = reserve_data(rhs.data_, rhs.is_altrep_, rhs.length_);
+  protect_ = detail::store::insert(data_);
+  is_altrep_ = ALTREP(data_);
+  data_p_ = get_p(is_altrep_, data_);
+  length_ = rhs.length_;
+  capacity_ = rhs.length_;
+}
```

Comment:
This avoids the double copy from https://github.com/r-lib/cpp11/issues/369

Before

```
> profmem::profmem(test_())
Rprofmem memory profiling of:
test_()

Memory allocations:
       what   bytes              calls
1     alloc  400048 test_() -> .Call()
2     alloc  800048 test_() -> .Call()
3     alloc  400056 test_() -> .Call()
4     alloc  400056 test_() -> .Call() # <- this one is extra
total       2000208   
```

After 

```
> profmem::profmem(test_())
Rprofmem memory profiling of:
test_()

Memory allocations:
       what   bytes              calls
1     alloc  400048 test_() -> .Call()
2     alloc  800048 test_() -> .Call()
3     alloc  400056 test_() -> .Call()
total       1600152  
```